File: Blackjack.java
import java.util.Scanner;
import java.util.ArrayList;

public class Blackjack extends CardGame {

    // フィールド定義
    private ArrayList<Card> playerDeck = new ArrayList<Card>();
    private ArrayList<Card> dealerDeck = new ArrayList<Card>();
    private int addCardTimes = 0;
    private int addScore = 0;
    private int turn = 0;
    private boolean playerTurn = true;
    private int playerScore = 0;
    private int dealerScore = 0;

    // コンストラクタ
    public Blackjack() {
        super(true);
    }

    // playメソッド
    public void play() {
        Scanner scan = new Scanner(System.in);

        addCard("both");
        addCard("both");

        System.out.println("ゲームスタート!");
        System.out.println("ディーラーのオープンカードは " + dealerDeck.get(0) + " です。");
        System.out.println("");

        LOOP: while (playerScore <= 21) {
            System.out.println("あなたのカードは");
            System.out.println(playerDeck.get(0) + " : " + cardScore(playerDeck.get(0).number));
            playerScore += addScore;
            System.out.println(playerDeck.get(1) + " : " + cardScore(playerDeck.get(1).number));
            playerScore += addScore;

            // 追加したカードの文字列表示
            if (turn >= 1) {
                addCard("player");
                for (int i = 2; i <= turn + 1; i++) {
                    System.out.println(playerDeck.get(i) + " : " + cardScore(playerDeck.get(i).number));
                    playerScore += addScore;
                }
            }

            System.out.println("(合計: " + playerScore + "点)");
            System.out.println("です。");
            System.out.println("");

            // playerが21点を超えた場合
            if (playerScore > 21) {
                System.out.println("21点を越えたので失格（バスト）です。");
                System.out.println("");
                System.out.println("ゲームに負けました...。");
                break;
            }

            // カード追加の選択
            while (true) {
                System.out.print("カードを追加（ヒット）しますか?(y|n) ");
                String answer = scan.nextLine();
                if (answer.equals("y")) {
                    break;
                } else if (answer.equals("n")) {
                    match();
                    break LOOP;
                }
            }

            System.out.println("");
            turn += 1;
            playerScore = 0;
        }

        scan.close();

    }

    // カード追加
    private void addCard(String distributed) {
        // playerのみカード追加
        if (distributed.equals("player")) {
            playerDeck.add(deck[addCardTimes]);
            addCardTimes += 1;
        } else if (distributed.equals("dealer")) {
            // ディーラーのみ追加
            dealerDeck.add(deck[addCardTimes]);
            addCardTimes += 1;
        } else if (distributed.equals("both")) {
            // 両方のカード追加
            playerDeck.add(deck[addCardTimes]);
            addCardTimes += 1;
            dealerDeck.add(deck[addCardTimes]);
            addCardTimes += 1;
        }
    }

    // カードの得点定義
    private String cardScore(int number) {
        if (number >= 2 && number <= 9) {
            addScore = number;
            return number + "点";
        } else if (number >= 10 && number <= 13) {
            addScore = 10;
            return "10点";
        } else if (number == 1) {
            if (playerScore <= 10 && playerTurn) {
                addScore = 11;
            } else if (playerScore > 10 && playerTurn) {
                addScore = 1;
            } else if (dealerScore <= 10 && playerTurn == false) {
                addScore = 11;
            } else if (dealerScore > 10 && playerTurn == false) {
                addScore = 1;
            }
            return "11点(1点)";
        } else {
            addScore = 0;
            return "0点";
        }
    }

    // ディーラーのカードオープン
    private void match() {
        playerTurn = false;
        System.out.println("");
        System.out.println("ディーラーのカードは");
        System.out.println(dealerDeck.get(0) + " : " + cardScore(dealerDeck.get(0).number));
        dealerScore += addScore;
        System.out.println(dealerDeck.get(1) + " : " + cardScore(dealerDeck.get(1).number));
        dealerScore += addScore;

        // ディーラーの得点が17点を越えるまでカード追加
        if (dealerScore < 17) {
            for (int i = 2; i < deck.length; i++) {
                addCard("dealer");
                System.out.println(dealerDeck.get(i) + " : " + cardScore(dealerDeck.get(i).number));
                dealerScore += addScore;

                if (dealerScore >= 17) {
                    break;
                }
            }
        }

        System.out.println("(合計: " + dealerScore + "点)");
        System.out.println("です。");
        System.out.println("");

        // 勝敗判定
        if (dealerScore > 21) {
            System.out.println("21点を越えたので失格（バスト）です。");
            System.out.println("");
            System.out.println("ゲームに勝ちました!");
        } else if (playerScore > dealerScore) {
            System.out.println("ゲームに勝ちました!");
        } else if (playerScore < dealerScore) {
            System.out.println("ゲームに負けました...。");
        } else {
            System.out.println("ゲームは引き分けです。");
        }
    }
}

File: CardGameServer/src/main/resources/static/blackjack.js
class Blackjack extends CardGame {
  constructor() {
    super(true);
    this.dealerScoreArea = document.getElementById("blackjack-dealerScore");
    this.dealerCard = document.getElementById("blackjack-dealerCard");
    this.blackjackText = document.getElementById("blackjack-text");
    this.playerCard = document.getElementById("blackjack-playerCard");
    this.playerScoreArea = document.getElementById("blackjack-playerScore");
    this.btnHit = document.getElementById("btn-hit");
    this.btnStay = document.getElementById("btn-stay");
    this.btnHit.addEventListener("click", (event) => this.handlerBtnHit(event));
    this.btnStay.addEventListener("click", (event) =>
      this.handlerBtnStay(event)
    );
    this.playerDeck = new Array();
    this.dealerDeck = new Array();
    this.playerScore = 0;
    this.dealerScore = 0;
    this.aceTimes = 0; //1が出た回数
    this.hitClickTimes = 1;
    this.play();
  }

  //playメソッド
  play() {
    this.addCard("both");
    this.addCard("both");

    this.addScore(this.playerDeck[0].number, "player");
    this.addScore(this.playerDeck[1].number, "player");

    //最初に1が2回出た時
    if (this.aceTimes == 2) {
      this.playerScore -= 10;
      this.aceTimes -= 1;
    }

    //カードとスコア表示
    this.dealerCard.append(this.dealerDeck[0]);
    this.playerCard.append(this.playerDeck[0]);
    this.playerCard.append(this.playerDeck[1]);
    this.playerScoreArea.innerHTML = "Player : " + this.playerScore;
  }

  addCard(distributed) {
    // playerのみカード追加
    if (distributed === "player") {
      this.playerDeck.push(this.deck.shift());
    }
    // ディーラーのみ追加
    else if (distributed === "dealer") {
      this.dealerDeck.push(this.deck.shift());
    }
    // 両方のカード追加
    else if (distributed === "both") {
      this.playerDeck.push(this.deck.shift());
      this.dealerDeck.push(this.deck.shift());
    }
  }

  //得点計算方法
  addScore(num, player) {
    //カードの数字が1の時
    if (num == 1) {
      this.aceTimes += 1;
      if (player === "player") {
        this.playerScore += 11;
      } else if (player === "dealer") {
        this.dealerScore += 11;
      }
    }
    //カードの数字が2~9の時
    else if (num >= 2 && num <= 9) {
      if (player === "player") {
        this.playerScore += num;
      } else if (player === "dealer") {
        this.dealerScore += num;
      }
    }
    //カードの数字が10~13の時
    else if (num >= 10 && num <= 13) {
      if (player === "player") {
        this.playerScore += 10;
      } else if (player === "dealer") {
        this.dealerScore += 10;
      }
    }
  }

  //ヒットボタンが押された時
  handlerBtnHit(event) {
    this.hitClickTimes += 1;
    this.addCard("player");
    this.addScore(this.playerDeck[this.hitClickTimes].number, "player");
    this.checkAce("player");

    //バーストした時
    if (this.playerScore > 21) {
      document.getElementById("blackjack-btn").style.display = "none";
      this.blackjackText.innerHTML = "あなたの負けです。";
    }

    //カードとスコア表示
    this.playerCard.append(this.playerDeck[this.hitClickTimes]);
    this.playerScoreArea.innerHTML = "Player : " + this.playerScore;
  }

  //ステイボタンがクリックされた時
  handlerBtnStay(event) {
    //ボタン非表示
    document.getElementById("blackjack-btn").style.display = "none";

    //一度aceTimesをリセット
    this.aceTimes = 0;

    this.addScore(this.dealerDeck[0].number, "dealer");
    this.addScore(this.dealerDeck[1].number, "dealer");
    this.dealerCard.append(this.dealerDeck[1]);

    //最初に1が2回出た場合
    if (this.aceTimes == 2) {
      this.dealerScore -= 10;
      this.aceTimes -= 1;
    }

    //ディーラーのスコアを17以上にする
    if (this.dealerScore < 17) {
      for (let i = 2; i < this.deck.length; i++) {
        this.addCard("dealer");
        this.addScore(this.dealerDeck[i].number, "dealer");
        this.dealerCard.append(this.dealerDeck[i]);
        this.checkAce("dealer");
        if (this.dealerScore >= 17) {
          break;
        }
      }
    }

    //勝敗判定
    if (this.dealerScore > 21) {
      this.blackjackText.innerHTML = "あなたの勝ちです。";
    } else if (this.playerScore > this.dealerScore) {
      this.blackjackText.innerHTML = "あなたの勝ちです。";
    } else if (this.playerScore < this.dealerScore) {
      this.blackjackText.innerHTML = "あなたの負けです。";
    } else {
      this.blackjackText.innerHTML = "引き分けです。";
    }

    //スコア表示
    this.dealerScoreArea.innerHTML = "Dealer : " + this.dealerScore;
  }

  //デッキに1がある時の再計上処理
  checkAce(checkPlayer) {
    switch (checkPlayer) {
      case "player": {
        if (this.playerScore > 21 && this.aceTimes > 0) {
          while (this.playerScore > 21) {
            this.playerScore -= 10;
            this.aceTimes -= 1;
            if (this.aceTimes <= 0) {
              break;
            }
          }
        }
        break;
      }
      case "dealer": {
        if (this.dealerScore > 21 && this.aceTimes > 0) {
          while (this.dealerScore > 21) {
            this.dealerScore -= 10;
            this.aceTimes -= 1;
            if (this.aceTimes <= 0) {
              break;
            }
          }
        }
        break;
      }
      default: {
        break;
      }
    }
  }
}
